source/game/field/obj/ObjectAurora.cc

```diff
@@ -0,0 +1,314 @@
+bool ObjectAurora::checkSphereFullImpl(f32 radius, const EGG::Vector3f &v0,
+        const EGG::Vector3f & /*v1*/, KCLTypeMask flags, CollisionInfo *pInfo,
+        KCLTypeMask *pFlagsOut, u32 timeOffset, bool push) {
+    EGG::Vector3f vel = v0 - m_pos;
+
+    if (vel.z < 0.0f || vel.z > COLLISION_SIZE.z || EGG::Mathf::abs(vel.x) > COLLISION_SIZE.x) {
+        return false;
+    }
+
+    if (!(flags & KCL_TYPE_FLOOR)) {
+        return false;
+    }
+
+    u32 t = timeOffset + System::RaceManager::Instance()->timer();
+    EGG::Vector3f bbox;
+    EGG::Vector3f fnrm;
+    f32 dist;
+
+    if (!calcCollision(radius, vel, t, bbox, fnrm, dist)) {
+        return false;
+    }
+
+    if (pInfo) {
+        pInfo->bbox.min = pInfo->bbox.min.minimize(bbox);
+        pInfo->bbox.max = pInfo->bbox.max.maximize(bbox);
+
+        pInfo->updateFloor(dist, fnrm);
+    }
+
+    if (pFlagsOut) {
+        auto *colDirector = CollisionDirector::Instance();
+
+        if (push) {
+            colDirector->pushCollisionEntry(dist, pFlagsOut, KCL_TYPE_BIT(COL_TYPE_ROAD2),
+                    COL_TYPE_ROAD2);
+            colDirector->setCurrentCollisionVariant(7);
+            colDirector->setCurrentCollisionTrickable(true);
+        } else {
+            *pFlagsOut |= KCL_TYPE_BIT(COL_TYPE_ROTATING_ROAD);
+        }
+
+        if (vel.z > COLLISION_SIZE.z - 600.0f * 4.0f) {
+            if (push) {
+                colDirector->pushCollisionEntry(dist, pFlagsOut, KCL_TYPE_BIT(COL_TYPE_BOOST_RAMP),
+                        COL_TYPE_BOOST_RAMP);
+            } else {
+                *pFlagsOut |= KCL_TYPE_BIT(COL_TYPE_ROAD2) | KCL_TYPE_BIT(COL_TYPE_BOOST_RAMP);
+            }
+        }
+    }
+
+    return true;
+}
+
+/// @addr{0x807FB060}
+/// @brief Calculates the sin-like collision of the wavy road.
+/// @details It seems to be modeled as a quadratic chirp with a starting frequency of 1 and a max
+/// frequency of 4. The minimum frequency of 1 occurs 30s into the race, and the maximum frequency
+/// occurs 2min 13s into the race. The "observed" frequency is dependent on the player's velocity
+/// towards the wavy road as well.
+bool ObjectAurora::calcCollision(f32 radius, const EGG::Vector3f &vel, u32 time, EGG::Vector3f &v0,
+        EGG::Vector3f &fnrm, f32 &dist) {
+    constexpr float INITIAL_FREQUENCY = 1.0f;
+    constexpr float MAX_FREQUENCY = 4.0f;
+    constexpr float PHASE_SHIFT_SECONDS = 30.0f;
+    constexpr float COLLISION_DISTANCE_THRESHOLD = 600.0f;
+    constexpr float UPWARP_THRESHOLD = 300.0f;
+    constexpr float UPWARP_DIST_SCALAR = 0.2f;
+
+    f32 minsNormalized = (static_cast<f32>(time) / 60.0f - PHASE_SHIFT_SECONDS) / 60.0f;
+    f32 frequency = std::min(INITIAL_FREQUENCY + minsNormalized * minsNormalized, MAX_FREQUENCY);
+    f32 velPeriod = (F_PI * (2.0f * vel.z)) / COLLISION_SIZE.z;
+
+    f32 result = EGG::Mathf::SinFIdx(
+            RAD2FIDX * (velPeriod * frequency + (F_PI * static_cast<f32>(time)) / 50.0f));
```

Comment:
```suggestion
            RAD2FIDX * (velPeriod * frequency + (F_PI * static_cast<f32>(time) / 50.0f)));
```
Can we double check that this is right?

Reply:
You're correct